File: core/src/main/java/com/xmartlabs/base/core/module/CoreAndroidModule.java
package com.xmartlabs.base.core.module;

import android.app.Application;
import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.xmartlabs.base.core.model.BuildInfo;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module
public class CoreAndroidModule {
  private final Application application;

  public CoreAndroidModule(Application application) {
    this.application = application;
  }

  @Provides
  @Singleton
  public Context provideApplicationContext() {
    return application;
  }

  @Provides
  @Singleton
  public BuildInfo provideBuildInformation() {
    return null;
  }

  @Provides
  @Singleton
  public SharedPreferences provideSharedPreferences() {
    return PreferenceManager.getDefaultSharedPreferences(application);
  }
}

File: core/src/main/java/com/xmartlabs/bigbang/core/module/CoreAndroidModule.kt
package com.xmartlabs.bigbang.core.module

import android.app.Application
import android.content.Context
import android.preference.PreferenceManager
import com.xmartlabs.bigbang.core.model.BuildInfo
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
open class CoreAndroidModule(private val application: Application) {
  @Provides
  @Singleton
  fun provideApplicationContext(): Context = application

  @Provides
  @Singleton
  open fun provideBuildInformation(): BuildInfo =
      throw IllegalStateException("A BuildInfo concrete class must be provided")

  @Provides
  @Singleton
  fun provideSharedPreferences() = PreferenceManager.getDefaultSharedPreferences(application)
}
